pkg/networkpolicy/networkpolicy.go

```diff
@@ -0,0 +1,274 @@
+func (a *NetworkPolicyAdvisor) localPodKey(e types.KubernetesConnectionEvent) (ret string) {
+	return e.LocalPodNamespace + ":" + a.labelKeyString(e.LocalPodLabels)
+}
+
+func (a *NetworkPolicyAdvisor) networkPeerKey(e types.KubernetesConnectionEvent) (ret string) {
+	if e.RemoteKind == "pod" {
+		ret = e.RemoteKind + ":" + e.RemotePodNamespace + ":" + a.labelKeyString(e.RemotePodLabels)
+	} else if e.RemoteKind == "svc" {
+		ret = e.RemoteKind + ":" + e.RemoteSvcNamespace + ":" + a.labelKeyString(e.RemoteSvcLabelSelector)
+	} else if e.RemoteKind == "other" {
+		ret = e.RemoteKind + ":" + e.RemoteOther
+	}
+	return ret + ":" + string(e.Port)
+}
+
+func (a *NetworkPolicyAdvisor) eventToRule(e types.KubernetesConnectionEvent) (ports []networkingv1.NetworkPolicyPort, peers []networkingv1.NetworkPolicyPeer) {
+	port := intstr.FromInt(int(e.Port))
+	protocol := v1.Protocol("TCP")
+	ports = []networkingv1.NetworkPolicyPort{
+		networkingv1.NetworkPolicyPort{
+			Port:     &port,
+			Protocol: &protocol,
+		},
+	}
+	// TODO: check if LocalPodNamespace != Remote*Namespace
+	if e.RemoteKind == "pod" {
+		peers = []networkingv1.NetworkPolicyPeer{
+			networkingv1.NetworkPolicyPeer{
+				PodSelector: &metav1.LabelSelector{MatchLabels: a.labelFilter(e.RemotePodLabels)},
+			},
+		}
+	} else if e.RemoteKind == "svc" {
+		peers = []networkingv1.NetworkPolicyPeer{
+			networkingv1.NetworkPolicyPeer{
+				PodSelector: &metav1.LabelSelector{MatchLabels: e.RemoteSvcLabelSelector},
+			},
+		}
+	} else if e.RemoteKind == "other" {
+		peers = []networkingv1.NetworkPolicyPeer{
+			networkingv1.NetworkPolicyPeer{
+				IPBlock: &networkingv1.IPBlock{
+					CIDR: e.RemoteOther + "/32",
+				},
+			},
+		}
+	} else {
+		panic("unknown event")
+	}
+	return
+}
+
+func (a *NetworkPolicyAdvisor) GeneratePolicies() {
+	eventsBySource := map[string][]types.KubernetesConnectionEvent{}
+	for _, e := range a.Events {
+		key := a.localPodKey(e)
+		if _, ok := eventsBySource[key]; ok {
+			eventsBySource[key] = append(eventsBySource[key], e)
+		} else {
+			eventsBySource[key] = []types.KubernetesConnectionEvent{e}
+		}
+	}
+
+	for _, events := range eventsBySource {
+		egressNetworkPeer := map[string][]types.KubernetesConnectionEvent{}
+		ingressNetworkPeer := map[string][]types.KubernetesConnectionEvent{}
+		for _, e := range events {
+			key := a.networkPeerKey(e)
+			if e.Type == "connect" {
+				if _, ok := egressNetworkPeer[key]; ok {
+					egressNetworkPeer[key] = append(egressNetworkPeer[key], e)
+				} else {
+					egressNetworkPeer[key] = []types.KubernetesConnectionEvent{e}
+				}
+			} else if e.Type == "accept" {
+				if _, ok := ingressNetworkPeer[key]; ok {
+					ingressNetworkPeer[key] = append(ingressNetworkPeer[key], e)
+				} else {
+					ingressNetworkPeer[key] = []types.KubernetesConnectionEvent{e}
+				}
+			}
+		}
+		egressPolicies := []networkingv1.NetworkPolicyEgressRule{}
+		for _, p := range egressNetworkPeer {
+			ports, peers := a.eventToRule(p[0])
```

Comment:
Above the events are saved as a list for each entry on the `egressNetworkPeer` map, and here only the first element of that list is used, so, why to save the full list?

Reply:
Indeed, only one item is necessary. The previous loop should be simplified to only save 1 element, or skip if there is already an item in `egressNetworkPeer[key]` or `ingressNetworkPeer[key]`.